api/envoy/config/core/v3/base.proto

```diff
@@ -323,6 +340,11 @@ message DataSource {
     // Local filesystem data source.
     string filename = 1 [(validate.rules).string = {min_len: 1}];
 
+    // Local filesystem data source watched for changes, e.g. by inotify on
+    // Linux. Not all *DataSource* instantiations support this, please consult
+    // documentation for the enclosing message.
```

Comment:
I'm not super happy with this contextual awareness or the need to have whoever is processing the parent message worry about concerns like atomic symlink resolution. An alternative idea I had was to allow an idea of virtual filesystems to be established similar to runtime and indirect into them from a DataSource. But, this is quite a bit more complicated and I'm not sure who really cares about atomicity here other than key rotation in the current API.

Reply:
Yeah agreed, wondering what the implementation would look like as well. The other question is how this is going to be validated.

Reply:
The implementation is actually super simple, if you take a look at https://github.com/envoyproxy/envoy/blob/9ff2d695ff12512fb7c838fd8b7afc3e853d3e0c/source/common/secret/sds_api.cc#L58, you can see that the watch is already computed based on gathering up all the `DataSource`s and computing paths, here we just need to do figure out common roots and subdirectories. That's kind of why I'm inclined towards this approach over essentially generalizing runtime, which is a much bigger project.

What would sway me towards a more general virtual filesystem would be use cases beyond key rotation where atomicity/snapshots matters. A quick grep reveals that this doesn't seem to be the case in the current API, but I'd be keen to hear on what we might benefit from in the future.

In terms of validation, I'd plan on basic unit/integration tests to validate the functional aspects. For the atomicity of swap, I think we could add a virtual function to `SdsApi` that the file reload code calls into after each file read, in the unit test we could override with something that inject delay and allows us to perform additional rotations.

Reply:
Should we still support `filename` above? Could everything be watched even if the user never actually moves it?

Reply:
This is an interesting idea, but I think the issue is that we need to add callbacks and behaviors at every site in Envoy that handles a `DataSource`. E.g. the LuaJIT filter, which might normally just be loading a file statically once. If you'd like to see this happen I can open a tracking issue for the various sites that would be missing this, but it's probably a fairly big task that requires per-subsystem integration and testing to get the refresh logic right. Even for TLS certs, the difficulty isn't watching files, it's dealing with dynamic updates of TLS on the listener objects (hence the SDS dynamic provider framework).

Reply:
Would there be any backward compatible way to replace `DataSource` with something like `WatchedDataSource` in those use cases that do support reloading?

Reply:
Yeah, that's a reasonable idea, we could deprecate the existing `DataSource` fields in the TLS certificate message, it's a bit of API churn but probably worth it from a clarity perspective.